### PR TITLE
[WIP] Standardizing datasets

### PR DESCRIPTION
**Pull request by k-chaney**
So this is my initial take on #242. There is a light wrapper around all of the datasets provided by torchvision which should be relatively easy to maintain.

`spike_encoders.py` houses wrappers around your encoding that are setup to work like the transforms provided by torchvision. Should this live in `bindsnet/datasets` or should it live in `bindsnet/encoding` since these are relatively generic objects?

The changes to `bindsnet/network/__init__.py` is motivated by keeping the default pytorch DataLoader which creates (in our case) BxTxCxHxW and having the time dimension be arbitrary seems like a good solution.

Not everything is implemented yet (hence WIP), but I wanted to make sure this was heading in a direction you guys are comfortable with before spending the time to wrap everything up. The TODO list includes putting spoken_mnist back in, example for a non-convolutional network, update other examples to use this new loading paradigm, work in a dataset that originated as spiking (simply to provide an example from that direction).

### CODE REVIEW

File: bindsnet/datasets/torchvision_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        def __init__(self,
+                     image_encoder: Encoder,
+                     label_encoder: Optional[Encoder],
+                     *args, **kwargs):
+            """
+            Constructor for the BindsNET torchvision dataset wrapper.
+            For details on the dataset you're interested in visit
+
+            https://pytorch.org/docs/stable/torchvision/datasets.html
+
```
Comment (djsaunde) [inline bindsnet/datasets/torchvision_wrapper.py]: Delete this line

File: bindsnet/network/__init__.py [hunk 2]
```diff
@@ -231,6 +231,8 @@ def run(self, inpts: Dict[str, torch.Tensor], time: int, **kwargs) -> None:
         :param Union[float, torch.Tensor] reward: Scalar value used in reward-modulated learning.
         :param Dict[Tuple[str], torch.Tensor] masks: Mapping of connection names to boolean masks determining which
                                                      weights to clamp to zero.
+        :param Union[int, Dict[str, int]] input_time_dim: Dimension for slicing in time for inputs. Can vary
```
Comment (djsaunde) [inline bindsnet/network/__init__.py]: Is this necessary? Can it just always be 0?

File: bindsnet/datasets/torchvision_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+from typing import Optional
+
+import torchvision
+
+from .spike_encoders import Encoder
+
+def vd_wrapper_creator(ds_type):
```
Comment (djsaunde) [inline bindsnet/datasets/torchvision_wrapper.py]: What does `vd` stand for?

File: bindsnet/datasets/__init__.py [hunk 4]
```diff
@@ -1,945 +1,9 @@
...
-                d['data'].append(temp[b'data'].reshape(-1, 3, 32, 32).transpose(0, 2, 3, 1))
-                d['labels'].append(temp[b'fine_labels'])
-
-        data, labels = np.concatenate(d['data']), np.concatenate(d['labels'])
-        return data.astype(np.float32), labels.astype(np.float32)
+MNIST = vd_wrapper_creator('MNIST')
+CIFAR10 = vd_wrapper_creator('CIFAR10')
+CIFAR100 = vd_wrapper_creator('CIFAR100')
+FashionMNIST = vd_wrapper_creator('FashionMNIST')
+ImageFolder = vd_wrapper_creator('ImageFolder')
```
Comment (djsaunde) [inline bindsnet/datasets/__init__.py]: Very cool. Can you add the rest of the PyTorch datasets?

File: bindsnet/datasets/__init__.py [hunk 5]
```diff
@@ -1,945 +1,9 @@
...
 
+from .spike_encoders import *
```
Comment (djsaunde) [inline bindsnet/datasets/__init__.py]: Why import everything here? Might want to do `__all__` = ['spike_encoder']`, which allows Python to find the submodule. See e.g. [here](https://github.com/Hananel-Hazan/bindsnet/blob/master/bindsnet/__init__.py).

File: bindsnet/datasets/spike_encoders.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
+import torch
+from typing import Tuple, List, Dict
+
+from .. import encoding
+
+__all__ = ['Encoder', 'SingleEncoder', 'RepeatEncoder',
+           'BernoulliEncoder', 'PoissonEncoder', 'RankOrderEncoder']
+
```
Comment (djsaunde) [inline bindsnet/datasets/spike_encoders.py]: Add line here

File: bindsnet/network/__init__.py [hunk 7]
```diff
@@ -283,7 +293,13 @@ def run(self, inpts: Dict[str, torch.Tensor], time: int, **kwargs) -> None:
             for l in self.layers:
                 # Update each layer of nodes.
                 if isinstance(self.layers[l], AbstractInput):
-                    self.layers[l].forward(x=inpts[l][t])
+                    # grab the time slice for each input
```
Comment (djsaunde) [inline bindsnet/network/__init__.py]: I'm not sure what these lines do, could you explain?

**Comment by djsaunde** [comment]
I really like the `Encoder` abstraction, and that it uses the already existing `bindsnet.encoding` functions. I'm interested to see how this works out!

File: bindsnet/network/__init__.py (hunk 2 above)
Comment (k-chaney) [reply bindsnet/network/__init__.py]: This was the simplest way I could think of to deal with the consequences of the pytorch DataLoader and default_collate_fn which will make everything BxTxCxWxH. Essentially we would have to maintain our own collate_fn and notation for which tensors should have the batch as the first dimension and which shouldn't.

File: bindsnet/datasets/torchvision_wrapper.py (hunk 3 above)
Comment (k-chaney) [reply bindsnet/datasets/torchvision_wrapper.py]: VisionDataset -- this is something that they're working on adding. I was reading their source code when putting this together and not their docs. I'll switch it to something more meaningful for the current implementation in torchvision.

File: bindsnet/datasets/__init__.py (hunk 4 above)
Comment (k-chaney) [reply bindsnet/datasets/__init__.py]: Yes, I'll add the rest, just wanted a small subset to test with initially. Should I make a note for which ones I've tested and which ones I haven't? Some I won't be able to (e.g. cityscapes and other large ones)

File: bindsnet/datasets/__init__.py (hunk 5 above)
Comment (k-chaney) [reply bindsnet/datasets/__init__.py]: I can change it to that.

File: bindsnet/datasets/__init__.py (hunk 4 above)
Comment (Hananel-Hazan) [reply bindsnet/datasets/__init__.py]: That is a good idea

Comment (djsaunde) [reply bindsnet/datasets/__init__.py]: Yes, could you please create a `README.md` in the `datasets/` folder which has these notes?

File: bindsnet/network/__init__.py (hunk 7 above)
Comment (k-chaney) [reply bindsnet/network/__init__.py]: Essentially, to deal with the fact that the time dimension could be anywhere I keep around a dictionary of slices for each input separately (different shape, time dimension, etc.) and I fill in the time into that list at the appropriate dimension. `slice(None)` is the same as `:`. I hope that helps clarify a bit.

Comment (djsaunde) [reply bindsnet/network/__init__.py]: Yup, thanks.

**Comment by k-chaney** [comment]
So I've been working on this more and this has pretty far reaching consequences inside of Pipeline, Environment, and all of the examples. I will keep moving on this, my only question revolves around backwards compatibility. There are some locations that have pretty hard dependencies to maintain long term such as:

https://github.com/Hananel-Hazan/bindsnet/blob/4021360cc0f715947bf6dc3f7effab003141cb4c/bindsnet/environment/__init__.py#L169

https://github.com/Hananel-Hazan/bindsnet/blob/4021360cc0f715947bf6dc3f7effab003141cb4c/bindsnet/environment/__init__.py#L260

https://github.com/Hananel-Hazan/bindsnet/blob/4021360cc0f715947bf6dc3f7effab003141cb4c/bindsnet/pipeline/__init__.py#L214


@djsaunde What're your thoughts regarding the overall impact and if I should put a lot of effort into maintaining backwards compatibility? My plan as of right now is to create more generic base classes and then provide examples of sub classing them. This would wind up breaking current users' code to some degree, but it should be relatively small changes on their end.

**Comment by djsaunde** [comment]
> What're your thoughts regarding the overall impact and if I should put a lot of effort into maintaining backwards compatibility?

The hard dependencies that you mentioned we definitely don't want in the library. I'd say, feel free to break backwards compatibility here. If you can come up with an abstraction that will allow us to render datasets / environments using the `Pipeline` object, that would be great; having to write custom code for each is not a good approach.

**Comment by k-chaney** [comment]
@djsaunde I want to check on the function of the history component inside of the pipeline as I've been racking my brain for a little bit on this. It looks like it overwrites self.obs before passing it into the encoder to use a delta from a previous time as an observation. This seems like it would be a dataset or environment dependent component.

**Comment by djsaunde** [comment]
I'm on vacation at the moment; @dee0512 might be good to ask about this!

**Comment by dee0512** [comment]
@k-chaney I just checked the pipeline. I agree, it should be part of dataset transformation rather than part of the pipeline if that is what you are asking.

**Comment by k-chaney** [comment]
@dee0512 Yes, that's what I was asking about. Everything is approaching a good state at this point. I just need to reintegrate the plotting utilities into the pipeline, add comments, and update examples (how many of the examples should I port?). Would you mind doing a quick read through to make sure everything makes sense?

**Comment by djsaunde** [comment]
@k-chaney I'm back from vacation; hopefully I'll have some time today to do a read-through and add some comments.

File: bindsnet/environment/__init__.py [hunk 8]
```diff
@@ -126,12 +126,22 @@ def step(self, a: int) -> Tuple[torch.Tensor, float, bool, Dict[Any, Any]]:
             self.update_history()
             self.update_index()
 
-        self.obs = self.encoder(self.obs)
+        raw_obs = self.obs
+
+        # The new standard is BxTxCxHxW. The gym environment doesn't
+        # follow exactly the same protocol.
+        self.obs = self.encoder(self.obs).unsqueeze(1)
```
Comment (djsaunde) [inline bindsnet/environment/__init__.py]: Should this be `unsqueeze(0)`?

File: bindsnet/environment/__init__.py [hunk 9]
```diff
@@ -126,12 +126,22 @@ def step(self, a: int) -> Tuple[torch.Tensor, float, bool, Dict[Any, Any]]:
...
+        if self.obs.dim() == 4:
+            self.obs = self.obs.unsqueeze(2)
 
         self.episode_step_count += 1
 
         # Return converted observations and other information.
-        return self.obs, self.reward, self.done, info
+        return {"obs": self.obs,
```
Comment (djsaunde) [inline bindsnet/environment/__init__.py]: Why put this in a dictionary? This breaks the Gym environment standard

File: bindsnet/pipeline/environment_pipeline.py [hunk 10]
```diff
@@ -115,8 +117,10 @@ def env_step(self):
         # Accumulate reward
         self.accumulated_reward += self.reward
 
-        return {'obs': self.obs, 'reward': reward,
-                'done': self.done, 'info': info}
+        batch['reward'] = self.reward
```
Comment (djsaunde) [inline bindsnet/pipeline/environment_pipeline.py]: Should return obs (state), reward, done, info

File: bindsnet/environment/__init__.py (hunk 8 above)
Comment (k-chaney) [reply bindsnet/environment/__init__.py]: I've rewritten this part to be a bit more clear

File: bindsnet/environment/__init__.py (hunk 9 above)
Comment (k-chaney) [reply bindsnet/environment/__init__.py]: Okay, I read up on the Gym environment standard and am now adding extra useful debug info (such as the original gym observation) to the info dictionary.

File: bindsnet/environment/__init__.py [hunk 11]
```diff
@@ -121,27 +121,35 @@ def step(self, a: int) -> Tuple[torch.Tensor, float, bool, Dict[Any, Any]]:
...
+
         # Store frame of history and encode the inputs.
         if len(self.history) > 0:
             self.update_history()
             self.update_index()
-
-        raw_obs = self.obs
+            # add the delta observation into the info for debugging and
+            # display
+            info['delta_obs'] = self.obs
```
Comment (djsaunde) [inline bindsnet/environment/__init__.py]: Aren't you just storing `self.obs` twice?

File: bindsnet/environment/__init__.py [hunk 12]
```diff
@@ -121,27 +121,35 @@ def step(self, a: int) -> Tuple[torch.Tensor, float, bool, Dict[Any, Any]]:
...
 
         # The new standard is BxTxCxHxW. The gym environment doesn't
         # follow exactly the same protocol.
-        self.obs = self.encoder(self.obs).unsqueeze(1)
-        if self.obs.dim() == 4:
-            self.obs = self.obs.unsqueeze(2)
+
+        # we want CxHxW. If it is in HxW add in  the channel dimension
```
Comment (djsaunde) [inline bindsnet/environment/__init__.py]: What if the data is 1D?

File: bindsnet/pipeline/__init__.py [hunk 13]
```diff
@@ -1,8 +1,10 @@
 from .environment_pipeline import EnvironmentPipeline
 from .base_pipeline import BasePipeline
 from .dataloader_pipeline import DataLoaderPipeline, TorchVisionDatasetPipeline
-from .pipeline_analysis import PipelineAnalyzer
+
+from . import pipeline_analysis
```
Comment (djsaunde) [inline bindsnet/pipeline/__init__.py]: I don't think you need to import these; as long as they're in `__all__`

File: bindsnet/environment/__init__.py (hunk 11 above)
Comment (k-chaney) [reply bindsnet/environment/__init__.py]: I pulled the history code that was inside of Pipeline previously. It overwrites `self.obs` internally. So the second will be the delta observation instead of the raw.

File: bindsnet/environment/__init__.py (hunk 12 above)
Comment (k-chaney) [reply bindsnet/environment/__init__.py]: Updated the code to include comments on exact assumptions as well as adding a flag in the constructor to not automatically add the channel dimension for 2D observations

File: bindsnet/pipeline/__init__.py (hunk 13 above)
Comment (k-chaney) [reply bindsnet/pipeline/__init__.py]: I think that the latest commit will reflect what you're looking for here.

File: examples/mnist/conv_mnist.py [hunk 14]
```diff
@@ -23,15 +26,15 @@
...
-parser.add_argument("--n_train", type=int, default=60000)
+parser.add_argument("--n_epochs", type=int, default=1)
 parser.add_argument("--n_test", type=int, default=10000)
 parser.add_argument("--kernel_size", type=int, default=16)
 parser.add_argument("--stride", type=int, default=4)
 parser.add_argument("--n_filters", type=int, default=25)
 parser.add_argument("--padding", type=int, default=0)
 parser.add_argument("--time", type=int, default=50)
 parser.add_argument("--dt", type=int, default=1.0)
-parser.add_argument("--intensity", type=float, default=0.5)
```
Comment (Hananel-Hazan) [inline examples/mnist/conv_mnist.py]: Is there is a reason why to change the default values?

File: examples/mnist/conv_mnist.py [hunk 15]
```diff
@@ -23,15 +26,15 @@
 
 parser = argparse.ArgumentParser()
 parser.add_argument("--seed", type=int, default=0)
-parser.add_argument("--n_train", type=int, default=60000)
+parser.add_argument("--n_epochs", type=int, default=1)
```
Comment (Hananel-Hazan) [inline examples/mnist/conv_mnist.py]: Is this should be 60000 like in the `--n_train`?

Comment (k-chaney) [reply examples/mnist/conv_mnist.py]: `--n_train` controls how many iterations. `--n_epochs` will control how many times through the dataset. Which for mnist (if using the standard train and test split) it will be 50k training steps for every epoch.

File: examples/mnist/conv_mnist.py (hunk 14 above)
Comment (k-chaney) [reply examples/mnist/conv_mnist.py]: Bindsnet's previous implementation of MNIST provided values 0-255 while Pytorch's provides values 0-1. Hence the change from intensity 0.5 to 128, so that the values fed into the PoissonEncoder is the same.

Comment (Hananel-Hazan) [reply examples/mnist/conv_mnist.py]: Thanks

File: examples/mnist/conv_mnist.py (hunk 15 above)
Comment (djsaunde) [reply examples/mnist/conv_mnist.py]: 60K, but I like this.

**Comment by djsaunde** [comment]
This looks like it's reaching a stable state. @k-chaney: Any more changes to be made before I make one last pass?

**Comment by k-chaney** [comment]
@djsaunde I don't have any more large changes for this pull request

**Comment by Hananel-Hazan** [comment]
Thank you @k-chaney for the Great work!
